File: mmupnp/src/main/java/net/mm2d/upnp/internal/impl/ControlPointImpl.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.impl;

import net.mm2d.log.Logger;
import net.mm2d.upnp.ControlPoint;
import net.mm2d.upnp.Device;
import net.mm2d.upnp.HttpClient;
import net.mm2d.upnp.IconFilter;
import net.mm2d.upnp.Protocol;
import net.mm2d.upnp.Service;
import net.mm2d.upnp.SsdpMessage;
import net.mm2d.upnp.SsdpMessageFilter;
import net.mm2d.upnp.internal.manager.DeviceHolder;
import net.mm2d.upnp.internal.manager.SubscribeManager;
import net.mm2d.upnp.internal.message.FakeSsdpMessage;
import net.mm2d.upnp.internal.parser.DeviceParser;
import net.mm2d.upnp.internal.server.SsdpNotifyReceiverList;
import net.mm2d.upnp.internal.server.SsdpSearchServerList;
import net.mm2d.upnp.internal.thread.TaskExecutors;
import net.mm2d.upnp.util.TextUtils;

import org.xml.sax.SAXException;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.xml.parsers.ParserConfigurationException;

/**
 * ControlPointの実装。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class ControlPointImpl implements ControlPoint {
    @Nonnull
    private final Protocol mProtocol;
    @Nonnull
    private SsdpMessageFilter mSsdpMessageFilter = SsdpMessageFilter.DEFAULT;
    @Nonnull
    private IconFilter mIconFilter = IconFilter.NONE;
    @Nonnull
    private final DiscoveryListenerList mDiscoveryListenerList;
    @Nonnull
    private final NotifyEventListenerList mNotifyEventListenerList;
    @Nonnull
    private final SsdpSearchServerList mSearchServerList;
    @Nonnull
    private final SsdpNotifyReceiverList mNotifyReceiverList;
    @Nonnull
    private final Map<String, DeviceImpl.Builder> mLoadingDeviceMap;
    @Nonnull
    private final Set<String> mEmbeddedUdnSet = new HashSet<>();
    @Nonnull
    private final TaskExecutors mTaskExecutors;
    @Nonnull
    private final AtomicBoolean mInitialized = new AtomicBoolean();
    @Nonnull
    private final AtomicBoolean mStarted = new AtomicBoolean();
    @Nonnull
    private final DeviceHolder mDeviceHolder;
    @Nonnull
    private final SubscribeManager mSubscribeManager;
    @Nonnull
    private final List<DeviceImpl.Builder> mLoadingPinnedDevices = Collections.synchronizedList(new ArrayList<>());

    public ControlPointImpl(
            @Nonnull final Protocol protocol,
            @Nonnull final Collection<NetworkInterface> interfaces,
            final boolean notifySegmentCheckEnabled,
            @Nonnull final DiFactory factory) {
        if (interfaces.isEmpty()) {
            throw new IllegalStateException("no valid network interface.");
        }
        mProtocol = protocol;
        mTaskExecutors = factory.createTaskExecutors();
        mLoadingDeviceMap = factory.createLoadingDeviceMap();
        mDiscoveryListenerList = new DiscoveryListenerList();
        mNotifyEventListenerList = new NotifyEventListenerList();

        mSearchServerList = factory.createSsdpSearchServerList(mTaskExecutors, interfaces, message ->
                mTaskExecutors.io(() -> onReceiveSsdpMessage(message)));
        mNotifyReceiverList = factory.createSsdpNotifyReceiverList(mTaskExecutors, interfaces, message ->
                mTaskExecutors.io(() -> onReceiveSsdpMessage(message)));
        mNotifyReceiverList.setSegmentCheckEnabled(notifySegmentCheckEnabled);
        mDeviceHolder = factory.createDeviceHolder(mTaskExecutors, this::lostDevice);
        mSubscribeManager = factory.createSubscribeManager(mTaskExecutors, mNotifyEventListenerList);
    }

    // VisibleForTesting
    @Nonnull
    HttpClient createHttpClient() {
        return new HttpClient(true);
    }

    // VisibleForTesting
    boolean needToUpdateSsdpMessage(
            @Nonnull final SsdpMessage oldMessage,
            @Nonnull final SsdpMessage newMessage) {
        final InetAddress newAddress = newMessage.getLocalAddress();
        if (mProtocol == Protocol.IP_V4_ONLY) {
            return newAddress instanceof Inet4Address;
        }
        if (mProtocol == Protocol.IP_V6_ONLY) {
            return newAddress instanceof Inet6Address;
        }
        final InetAddress oldAddress = oldMessage.getLocalAddress();
        if (oldAddress instanceof Inet4Address) {
            if (oldAddress.isLinkLocalAddress()) {
                return true;
            }
            return !(newAddress instanceof Inet6Address);
        } else {
            if (newAddress instanceof Inet6Address) {
                return true;
            }
            return newAddress != null && !newAddress.isLinkLocalAddress();
        }
    }

    // VisibleForTesting
    void onReceiveSsdpMessage(@Nonnull final SsdpMessage message) {
        if (mSsdpMessageFilter.accept(message)) {
            onAcceptSsdpMessage(message);
        }
    }

    // VisibleForTesting
    void onAcceptSsdpMessage(@Nonnull final SsdpMessage message) {
        synchronized (mDeviceHolder) {
            final String uuid = message.getUuid();
            final Device device = mDeviceHolder.get(uuid);
            if (device == null) {
                if (mEmbeddedUdnSet.contains(uuid)) {
                    return;
                }
                onReceiveNewSsdp(message);
                return;
            }
            if (TextUtils.equals(message.getNts(), SsdpMessage.SSDP_BYEBYE)) {
                if (!isPinnedDevice(device)) {
                    lostDevice(device);
                }
            } else {
                if (needToUpdateSsdpMessage(device.getSsdpMessage(), message)) {
                    device.updateSsdpMessage(message);
                }
            }
        }
    }

    private void onReceiveNewSsdp(@Nonnull final SsdpMessage message) {
        final String uuid = message.getUuid();
        if (TextUtils.equals(message.getNts(), SsdpMessage.SSDP_BYEBYE)) {
            mLoadingDeviceMap.remove(uuid);
            return;
        }
        final DeviceImpl.Builder deviceBuilder = mLoadingDeviceMap.get(uuid);
        if (deviceBuilder != null) {
            if (needToUpdateSsdpMessage(deviceBuilder.getSsdpMessage(), message)) {
                deviceBuilder.updateSsdpMessage(message);
            }
            return;
        }
        loadDevice(uuid, new DeviceImpl.Builder(this, mSubscribeManager, message));
    }

    private void loadDevice(
            @Nonnull final String uuid,
            @Nonnull final DeviceImpl.Builder builder) {
        mLoadingDeviceMap.put(uuid, builder);
        if (!mTaskExecutors.io(() -> loadDevice(builder))) {
            mLoadingDeviceMap.remove(uuid);
        }
    }

    private void loadDevice(@Nonnull final DeviceImpl.Builder builder) {
        final HttpClient client = createHttpClient();
        final String uuid = builder.getUuid();
        try {
            DeviceParser.loadDescription(client, builder);
            final Device device = builder.build();
            device.loadIconBinary(client, mIconFilter);
            synchronized (mDeviceHolder) {
                if (mLoadingDeviceMap.remove(uuid) != null) {
                    discoverDevice(device);
                }
            }
        } catch (final IOException | SAXException | ParserConfigurationException | RuntimeException e) {
            Logger.w(e);
            Logger.i(() -> e.getClass().getSimpleName() + " occurred on loadDevice\n" + builder.toDumpString());
            synchronized (mDeviceHolder) {
                mLoadingDeviceMap.remove(uuid);
            }
        } finally {
            client.close();
        }
    }

    @Override
    public void initialize() {
        if (mInitialized.getAndSet(true)) {
            return;
        }
        mDeviceHolder.start();
        mSubscribeManager.initialize();
    }

    @Override
    public void terminate() {
        if (mStarted.get()) {
            stop();
        }
        if (!mInitialized.getAndSet(false)) {
            return;
        }
        mTaskExecutors.terminate();
        mSubscribeManager.terminate();
        mDeviceHolder.shutdownRequest();
    }

    @Override
    public void start() {
        if (!mInitialized.get()) {
            initialize();
        }
        if (mStarted.getAndSet(true)) {
            return;
        }
        mSubscribeManager.start();
        mSearchServerList.openAndStart();
        mNotifyReceiverList.openAndStart();
    }

    @Override
    public void stop() {
        if (!mStarted.getAndSet(false)) {
            return;
        }
        mSubscribeManager.stop();
        mSearchServerList.stop();
        mNotifyReceiverList.stop();
        mSearchServerList.close();
        mNotifyReceiverList.close();
        final List<Device> list = getDeviceList();
        for (final Device device : list) {
            lostDevice(device);
        }
        mDeviceHolder.clear();
    }

    @Override
    public void clearDeviceList() {
        synchronized (mDeviceHolder) {
            final List<Device> list = getDeviceList();
            for (final Device device : list) {
                lostDevice(device);
            }
        }
    }

    @Override
    public void search() {
        search(null);
    }

    @Override
    public void search(@Nullable final String st) {
        if (!mStarted.get()) {
            throw new IllegalStateException("ControlPoint is not started.");
        }
        mSearchServerList.search(st);
    }

    @Override
    public void setSsdpMessageFilter(@Nullable final SsdpMessageFilter filter) {
        mSsdpMessageFilter = filter != null ? filter : SsdpMessageFilter.DEFAULT;
    }

    @Override
    public void setIconFilter(@Nullable final IconFilter filter) {
        mIconFilter = filter != null ? filter : IconFilter.NONE;
    }

    @Override
    public void addDiscoveryListener(@Nonnull final DiscoveryListener listener) {
        mDiscoveryListenerList.add(listener);
    }

    @Override
    public void removeDiscoveryListener(@Nonnull final DiscoveryListener listener) {
        mDiscoveryListenerList.remove(listener);
    }

    @Override
    public void addNotifyEventListener(@Nonnull final NotifyEventListener listener) {
        mNotifyEventListenerList.add(listener);
    }

    @Override
    public void removeNotifyEventListener(@Nonnull final NotifyEventListener listener) {
        mNotifyEventListenerList.remove(listener);
    }

    // VisibleForTesting
    void discoverDevice(@Nonnull final Device device) {
        Logger.d(() -> "discoverDevice:[" + device.getFriendlyName() + "](" + device.getIpAddress() + ")");
        if (isPinnedDevice(mDeviceHolder.get(device.getUdn()))) {
            return;
        }
        mEmbeddedUdnSet.addAll(collectEmbeddedUdn(device));
        mDeviceHolder.add(device);
        mTaskExecutors.callback(() ->
                mDiscoveryListenerList.onDiscover(device));
    }

    // VisibleForTesting
    void lostDevice(@Nonnull final Device device) {
        Logger.d(() -> "lostDevice:[" + device.getFriendlyName() + "](" + device.getIpAddress() + ")");
        mEmbeddedUdnSet.removeAll(collectEmbeddedUdn(device));
        synchronized (mDeviceHolder) {
            final List<Service> list = device.getServiceList();
            for (final Service s : list) {
                mSubscribeManager.unregister(s);
            }
            mDeviceHolder.remove(device);
        }
        mTaskExecutors.callback(() ->
                mDiscoveryListenerList.onLost(device));
    }

    // VisibleForTesting
    public static Set<String> collectEmbeddedUdn(@Nonnull final Device device) {
        final List<Device> deviceList = device.getDeviceList();
        if (deviceList.isEmpty()) {
            return Collections.emptySet();
        }
        final Set<String> outSet = new HashSet<>();
        for (final Device d : deviceList) {
            collectEmbeddedUdn(d, outSet);
        }
        return outSet;
    }

    private static void collectEmbeddedUdn(
            @Nonnull final Device device,
            @Nonnull final Set<String> outSet) {
        outSet.add(device.getUdn());
        for (final Device d : device.getDeviceList()) {
            collectEmbeddedUdn(d, outSet);
        }
    }

    @Override
    public int getDeviceListSize() {
        return mDeviceHolder.size();
    }

    @Override
    @Nonnull
    public List<Device> getDeviceList() {
        return mDeviceHolder.getDeviceList();
    }

    @Override
    @Nullable
    public Device getDevice(@Nonnull final String udn) {
        return mDeviceHolder.get(udn);
    }

    @Override
    public void tryAddDevice(
            @Nonnull final String uuid,
            @Nonnull final String location) {
        for (final Device device : getDeviceList()) {
            if (TextUtils.equals(device.getLocation(), location)) {
                Logger.i(() -> "already added: " + location);
                return;
            }
        }
        for (final DeviceImpl.Builder builder : mLoadingDeviceMap.values()) {
            if (TextUtils.equals(builder.getLocation(), location)) {
                Logger.i(() -> "already loading: " + location);
                return;
            }
        }
        final FakeSsdpMessage message = new FakeSsdpMessage(location, uuid, false);
        loadDevice(uuid, new DeviceImpl.Builder(this, mSubscribeManager, message));
    }

    @Override
    public void tryAddPinnedDevice(@Nonnull final String location) {
        for (final Device device : getDeviceList()) {
            if (TextUtils.equals(device.getLocation(), location) && isPinnedDevice(device)) {
                Logger.i(() -> "already added: " + location);
                return;
            }
        }
        final DeviceImpl.Builder builder = new DeviceImpl.Builder(
                this, mSubscribeManager, new FakeSsdpMessage(location));
        mLoadingPinnedDevices.add(builder);
        mTaskExecutors.io(() -> loadPinnedDevice(builder));
    }

    private void loadPinnedDevice(@Nonnull final DeviceImpl.Builder builder) {
        final HttpClient client = createHttpClient();
        try {
            DeviceParser.loadDescription(client, builder);
            final Device device = builder.build();
            device.loadIconBinary(client, mIconFilter);
            synchronized (mDeviceHolder) {
                final String udn = device.getUdn();
                if (!mLoadingPinnedDevices.remove(builder)) {
                    return;
                }
                mLoadingDeviceMap.remove(udn);
                final Device lostDevice = mDeviceHolder.remove(udn);
                if (lostDevice != null) {
                    lostDevice(lostDevice);
                }
                discoverDevice(device);
            }
        } catch (final IOException | IllegalStateException | SAXException | ParserConfigurationException e) {
            Logger.w(() -> "fail to load:\n" + builder.getLocation(), e);
        } finally {
            client.close();
        }
    }

    @Override
    public void removePinnedDevice(@Nonnull final String location) {
        synchronized (mLoadingPinnedDevices) {
            for (final ListIterator<DeviceImpl.Builder> i = mLoadingPinnedDevices.listIterator(); i.hasNext(); ) {
                final DeviceImpl.Builder builder = i.next();
                if (TextUtils.equals(builder.getLocation(), location)) {
                    i.remove();
                    return;
                }
            }
        }
        for (final Device device : getDeviceList()) {
            if (TextUtils.equals(device.getLocation(), location)) {
                lostDevice(device);
                return;
            }
        }
    }

    private static boolean isPinnedDevice(@Nullable final Device device) {
        return device != null && device.isPinned();
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/impl/ControlPointImpl.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.impl

import net.mm2d.upnp.Adapter.iconFilter
import net.mm2d.upnp.ControlPoint
import net.mm2d.upnp.ControlPoint.DiscoveryListener
import net.mm2d.upnp.ControlPoint.EventListener
import net.mm2d.upnp.ControlPoint.MulticastEventListener
import net.mm2d.upnp.ControlPoint.NotifyEventListener
import net.mm2d.upnp.Device
import net.mm2d.upnp.IconFilter
import net.mm2d.upnp.Protocol
import net.mm2d.upnp.Service
import net.mm2d.upnp.SingleHttpClient
import net.mm2d.upnp.SsdpMessage
import net.mm2d.upnp.internal.impl.DeviceImpl.Builder
import net.mm2d.upnp.internal.manager.DeviceHolder
import net.mm2d.upnp.internal.manager.SubscribeManager
import net.mm2d.upnp.internal.message.FakeSsdpMessage
import net.mm2d.upnp.internal.parser.DeviceParser
import net.mm2d.upnp.internal.server.DEFAULT_SSDP_MESSAGE_FILTER
import net.mm2d.upnp.internal.server.MulticastEventReceiverList
import net.mm2d.upnp.internal.server.SsdpNotifyServerList
import net.mm2d.upnp.internal.server.SsdpSearchServerList
import net.mm2d.upnp.internal.thread.TaskExecutors
import net.mm2d.upnp.internal.util.toSimpleTrace
import net.mm2d.upnp.log.Logger
import java.net.Inet4Address
import java.net.Inet6Address
import java.net.NetworkInterface
import java.util.*
import java.util.concurrent.CopyOnWriteArraySet
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Implements for [ControlPoint].
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class ControlPointImpl(
    private val protocol: Protocol,
    interfaces: Iterable<NetworkInterface>,
    notifySegmentCheckEnabled: Boolean,
    subscriptionEnabled: Boolean,
    multicastEventingEnabled: Boolean,
    factory: DiFactory
) : ControlPoint {
    private var iconFilter: IconFilter = EMPTY_FILTER
    private val discoveryListenerSet: MutableSet<DiscoveryListener>

    @Suppress("DEPRECATION")
    private val notifyEventListenerSet: MutableSet<NotifyEventListener>
    private val eventListenerSet: MutableSet<EventListener>
    private val multicastEventListenerSet: MutableSet<MulticastEventListener>
    private val searchServerList: SsdpSearchServerList
    private val notifyServerList: SsdpNotifyServerList
    private val deviceMap: MutableMap<String, Device>
    private val loadingDeviceMap: MutableMap<String, Builder>
    private val initialized = AtomicBoolean()
    private val started = AtomicBoolean()
    private val deviceHolder: DeviceHolder
    private val loadingPinnedDevices: MutableList<Builder>
    private val multicastEventReceiverList: MulticastEventReceiverList?
    internal val subscribeManager: SubscribeManager
    internal val taskExecutors: TaskExecutors

    init {
        check(interfaces.any()) { "no valid network interface." }
        discoveryListenerSet = CopyOnWriteArraySet()
        notifyEventListenerSet = CopyOnWriteArraySet()
        eventListenerSet = CopyOnWriteArraySet()
        multicastEventListenerSet = CopyOnWriteArraySet()
        deviceMap = mutableMapOf()
        loadingPinnedDevices = Collections.synchronizedList(mutableListOf())
        taskExecutors = factory.createTaskExecutors()
        loadingDeviceMap = factory.createLoadingDeviceMap()
        searchServerList = factory.createSsdpSearchServerList(taskExecutors, interfaces) { message ->
            taskExecutors.io { onReceiveSsdpMessage(message) }
        }
        notifyServerList = factory.createSsdpNotifyServerList(taskExecutors, interfaces) { message ->
            taskExecutors.io { onReceiveSsdpMessage(message) }
        }
        notifyServerList.setSegmentCheckEnabled(notifySegmentCheckEnabled)
        deviceHolder = factory.createDeviceHolder(taskExecutors) { lostDevice(it) }
        subscribeManager = factory.createSubscribeManager(subscriptionEnabled, taskExecutors, ::onReceiveEvent)
        multicastEventReceiverList = if (multicastEventingEnabled) {
            factory.createMulticastEventReceiverList(taskExecutors, interfaces, ::onReceiveMulticastEvent)
        } else null
    }

    private fun createHttpClient(): SingleHttpClient = SingleHttpClient.create(true)

    // VisibleForTesting
    internal fun needToUpdateSsdpMessage(oldMessage: SsdpMessage, newMessage: SsdpMessage): Boolean {
        val newAddress = newMessage.localAddress
        if (protocol === Protocol.IP_V4_ONLY) {
            return newAddress is Inet4Address
        }
        if (protocol === Protocol.IP_V6_ONLY) {
            return newAddress is Inet6Address
        }
        val oldAddress = oldMessage.localAddress
        return if (oldAddress is Inet4Address) {
            if (oldAddress.isLinkLocalAddress) true else newAddress is Inet4Address
        } else {
            if (newAddress is Inet6Address) true else newAddress?.isLinkLocalAddress == false
        }
    }

    // VisibleForTesting
    internal fun onReceiveSsdpMessage(message: SsdpMessage) {
        synchronized(deviceHolder) {
            val uuid = message.uuid
            val device = deviceMap[uuid]
            if (device == null) {
                onReceiveNewSsdp(message)
                return
            }
            if (message.nts == SsdpMessage.SSDP_BYEBYE) {
                if (!isPinnedDevice(device)) {
                    lostDevice(device)
                }
            } else {
                if (needToUpdateSsdpMessage(device.ssdpMessage, message)) {
                    device.updateSsdpMessage(message)
                }
            }
        }
    }

    private fun onReceiveNewSsdp(message: SsdpMessage) {
        val uuid = message.uuid
        if (message.nts == SsdpMessage.SSDP_BYEBYE) {
            loadingDeviceMap.remove(uuid)
            return
        }
        loadingDeviceMap[uuid]?.let {
            if (needToUpdateSsdpMessage(it.getSsdpMessage(), message)) {
                it.updateSsdpMessage(message)
            }
            return
        }
        loadDevice(uuid, Builder(this, message))
    }

    // VisibleForTesting
    internal fun loadDevice(uuid: String, builder: Builder) {
        loadingDeviceMap[uuid] = builder
        if (!taskExecutors.io { loadDevice(builder) }) {
            loadingDeviceMap.remove(uuid)
        }
    }

    private fun loadDevice(builder: Builder) {
        val client = createHttpClient()
        val uuid = builder.getUuid()
        try {
            DeviceParser.loadDescription(client, builder)
            val device = builder.build()
            device.loadIconBinary(client, iconFilter)
            synchronized(deviceHolder) {
                if (loadingDeviceMap.remove(uuid) != null) {
                    discoverDevice(device)
                }
            }
        } catch (e: Exception) {
            Logger.w { "loadDevice: " + e.toSimpleTrace() }
            Logger.i(e) { builder.toDumpString() }
            synchronized(deviceHolder) {
                loadingDeviceMap.remove(uuid)
            }
        } finally {
            client.close()
        }
    }

    private fun onReceiveEvent(service: Service, seq: Long, properties: List<Pair<String, String>>) {
        eventListenerSet.forEach {
            taskExecutors.callback { it.onEvent(service, seq, properties) }
        }
        if (notifyEventListenerSet.isEmpty()) return
        properties.forEach {
            notifyEvent(service, seq, it.first, it.second)
        }
    }

    private fun notifyEvent(service: Service, seq: Long, name: String?, value: String?) {
        val variable = service.findStateVariable(name)
        if (variable?.isSendEvents != true || value == null) {
            Logger.w { "illegal notify argument: $name $value" }
            return
        }
        notifyEventListenerSet.forEach {
            taskExecutors.callback { it.onNotifyEvent(service, seq, variable.name, value) }
        }
    }

    // VisibleForTesting
    internal fun onReceiveMulticastEvent(
        uuid: String,
        svcid: String,
        lvl: String,
        seq: Long,
        properties: List<Pair<String, String>>
    ) {
        val service = synchronized(deviceHolder) {
            deviceMap[uuid]?.findServiceById(svcid)
        } ?: return
        multicastEventListenerSet.forEach {
            taskExecutors.callback { it.onEvent(service, lvl, seq, properties) }
        }
    }

    override val deviceListSize: Int
        get() = deviceHolder.size

    override val deviceList: List<Device>
        get() = deviceHolder.deviceList

    override fun initialize() {
        if (initialized.getAndSet(true)) {
            return
        }
        deviceHolder.start()
        subscribeManager.initialize()
    }

    override fun terminate() {
        if (started.get()) {
            stop()
        }
        if (!initialized.getAndSet(false)) {
            return
        }
        taskExecutors.terminate()
        subscribeManager.terminate()
        deviceHolder.stop()
    }

    override fun start() {
        if (!initialized.get()) {
            initialize()
        }
        if (started.getAndSet(true)) {
            return
        }
        multicastEventReceiverList?.start()
        subscribeManager.start()
        searchServerList.start()
        notifyServerList.start()
    }

    override fun stop() {
        if (!started.getAndSet(false)) {
            return
        }
        multicastEventReceiverList?.stop()
        subscribeManager.stop()
        searchServerList.stop()
        notifyServerList.stop()
        deviceList.forEach { lostDevice(it) }
        deviceHolder.clear()
    }

    override fun clearDeviceList() {
        synchronized(deviceHolder) {
            deviceList.forEach { lostDevice(it) }
        }
    }

    override fun search(st: String?) {
        check(started.get()) { "ControlPoint is not started." }
        searchServerList.search(st)
    }

    override fun setSsdpMessageFilter(predicate: ((SsdpMessage) -> Boolean)?) {
        val predicateNonNull = predicate ?: DEFAULT_SSDP_MESSAGE_FILTER
        searchServerList.setFilter(predicateNonNull)
        notifyServerList.setFilter(predicateNonNull)
    }

    override fun setIconFilter(filter: IconFilter?) {
        iconFilter = filter ?: EMPTY_FILTER
    }

    override fun addDiscoveryListener(listener: DiscoveryListener) {
        discoveryListenerSet.add(listener)
    }

    override fun removeDiscoveryListener(listener: DiscoveryListener) {
        discoveryListenerSet.remove(listener)
    }

    @Suppress("DEPRECATION")
    override fun addNotifyEventListener(listener: NotifyEventListener) {
        notifyEventListenerSet.add(listener)
    }

    @Suppress("DEPRECATION")
    override fun removeNotifyEventListener(listener: NotifyEventListener) {
        notifyEventListenerSet.remove(listener)
    }

    override fun addEventListener(listener: EventListener) {
        eventListenerSet.add(listener)
    }

    override fun removeEventListener(listener: EventListener) {
        eventListenerSet.remove(listener)
    }

    override fun addMulticastEventListener(listener: MulticastEventListener) {
        multicastEventListenerSet.add(listener)
    }

    override fun removeMulticastEventListener(listener: MulticastEventListener) {
        multicastEventListenerSet.remove(listener)
    }

    // VisibleForTesting
    internal fun discoverDevice(device: Device) {
        Logger.d { "discoverDevice:[${device.friendlyName}](${device.ipAddress})" }
        if (isPinnedDevice(deviceHolder[device.udn])) {
            return
        }
        deviceHolder.add(device)
        collectUdn(device).forEach { deviceMap[it] = device }
        taskExecutors.callback {
            discoveryListenerSet.forEach { it.onDiscover(device) }
        }
    }

    // VisibleForTesting
    internal fun lostDevice(device: Device) {
        Logger.d { "lostDevice:[${device.friendlyName}](${device.ipAddress})" }
        synchronized(deviceHolder) {
            device.serviceList.forEach { subscribeManager.unregister(it) }
            collectUdn(device).forEach { deviceMap.remove(it) }
            deviceHolder.remove(device)
        }
        taskExecutors.callback {
            discoveryListenerSet.forEach { it.onLost(device) }
        }
    }

    override fun getDevice(udn: String): Device? = deviceMap[udn]

    override fun tryAddDevice(uuid: String, location: String) {
        if (deviceList.any { it.location == location }) {
            Logger.i { "already added: $location" }
            return
        }
        if (loadingDeviceMap.values.any { it.getLocation() == location }) {
            Logger.i { "already loading: $location" }
            return
        }
        val message = FakeSsdpMessage(location, uuid, false)
        loadDevice(uuid, Builder(this, message))
    }

    override fun tryAddPinnedDevice(location: String) {
        if (deviceList.any { it.location == location && isPinnedDevice(it) }) {
            Logger.i { "already added: $location" }
            return
        }
        val builder = Builder(this, FakeSsdpMessage(location))
        loadingPinnedDevices.add(builder)
        taskExecutors.io { loadPinnedDevice(builder) }
    }

    private fun loadPinnedDevice(builder: Builder) {
        val client = createHttpClient()
        try {
            DeviceParser.loadDescription(client, builder)
            val device = builder.build()
            device.loadIconBinary(client, iconFilter)
            synchronized(deviceHolder) {
                if (!loadingPinnedDevices.remove(builder)) {
                    return
                }
                val udn = device.udn
                loadingDeviceMap.remove(udn)
                deviceHolder.remove(udn)?.let { lostDevice(it) }
                discoverDevice(device)
            }
        } catch (e: Exception) {
            Logger.w(e) { "fail to load:\n${builder.getLocation()}" }
        } finally {
            client.close()
        }
    }

    override fun removePinnedDevice(location: String) {
        synchronized(loadingPinnedDevices) {
            val i = loadingPinnedDevices.listIterator()
            while (i.hasNext()) {
                if (i.next().getLocation() == location) {
                    i.remove()
                    return
                }
            }
        }
        deviceList.find { it.location == location }
            ?.let { lostDevice(it) }
    }

    companion object {
        private val EMPTY_FILTER = iconFilter { emptyList() }

        // VisibleForTesting
        internal fun collectUdn(device: Device): Set<String> = mutableSetOf<String>().also {
            collectUdn(device, it)
        }

        private fun collectUdn(device: Device, outSet: MutableSet<String>) {
            outSet.add(device.udn)
            device.deviceList.forEach { collectUdn(it, outSet) }
        }

        private fun isPinnedDevice(device: Device?): Boolean {
            return device?.isPinned == true
        }
    }
}
